io/include/pcl/io/real_sense_2_grabber.h

```diff
@@ -57,14 +57,19 @@ namespace pcl
   {
   public:
     /** \brief Constructor
-    * \param[in] serial_number optional to start up a specific device
     */
-    RealSense2Grabber (const uint32_t serial_number = 0);
+    RealSense2Grabber ();
 
-    /** \brief Constructor
-    * \param[in] file_name used for loading bag file
+    /** \brief Create with Bag file
+    * \param[in] file_name path to bag file
+    */
+    static RealSense2Grabber& createFromBagFile (const std::string& file_name) { return RealSense2Grabber (file_name, ""); } 
```

Comment:
This factory constructor will crash your app. You are returning a reference to a local variable that goes out of scope and is destructed as soon as the method returns.

Reply:
I have to admit that I'm not an expert at this but doesn't an optimized compiler know that the transferred reference should be copied over and stay in scope?

https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Named_Constructor

Reply:
Within the function the object is created on the stack. There is no technical possibility for it to stay in scope when the function returns. The reference is just a "glorified pointer", it can not copy any data, it only points (or refers) at some data.

In the link you provided the created object is returned by value (not by reference!). A good compiler will apply RVO in that case and we hopefully avoid unnecessary copy. But I would not rely on this, especially when we have some complex objects, threads, etc. as member fields.